File: app/src/main/java/com/example/constraintlayoutapp/MainActivity.java
package com.example.constraintlayoutapp;

import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintSet;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import android.os.Bundle;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import com.example.constraintlayoutapp.databinding.MainBinder;

public class MainActivity extends AppCompatActivity implements View.OnClickListener, View.OnTouchListener {

    private static final String TAG = MainActivity.class.getSimpleName();
    private MainBinder binder;
    private MainViewModel viewModel;
    private ConstraintSet constraintSet;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binder = DataBindingUtil.setContentView(this, R.layout.activity_main);
        viewModel = new ViewModelProvider(this).get(MainViewModel.class);

        binder.setViewModel(viewModel);
        binder.setLifecycleOwner(this);
        constraintSet = new ConstraintSet();

        binder.floatingActionButtons.floatingActionButtonAndroid.setOnClickListener(this::onClick);
        binder.floatingActionButtons.floatingActionButtonAndroid.setOnTouchListener(this::onTouch);

        binder.getViewModel().observeFABVisibility().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isShowed) {
                if (isShowed) showExtendedFAB();
                else hideExtendedFab();
            }
        });
    }

    @Override
    public void onClick(View view) {
        if (view == binder.floatingActionButtons.floatingActionButtonAndroid)
            binder.getViewModel().setLiveFABVisibility();
    }

    @Override
    public boolean onTouch(View view, MotionEvent motionEvent) {
        ViewGroup.MarginLayoutParams layoutParams = (ViewGroup.MarginLayoutParams) binder.floatingActionButtons.constraintLayout.getLayoutParams();
        switch (motionEvent.getAction()) {
            case MotionEvent.ACTION_DOWN:
                binder.getViewModel().setDownRawX(motionEvent.getRawX());
                binder.getViewModel().setDownRawY(motionEvent.getRawY());
                binder.getViewModel().setDx(binder.floatingActionButtons.constraintLayout.getX() - binder.getViewModel().getDownRawX());
                binder.getViewModel().setDy(binder.floatingActionButtons.constraintLayout.getY() - binder.getViewModel().getDownRawY());
                return false; //Ripple Effect
            case MotionEvent.ACTION_MOVE:
                binder.getViewModel().setViewDimension(binder.floatingActionButtons.constraintLayout);
                binder.getViewModel().setParentDimension(binder.floatingActionButtons.constraintLayout.getParent());
                binder.getViewModel().setNewX(motionEvent.getRawX(), layoutParams);
                binder.getViewModel().setNewY(motionEvent.getRawY(), layoutParams);
                binder.floatingActionButtons.constraintLayout.animate()
                        .x( binder.getViewModel().getNewX() )
                        .y( binder.getViewModel().getNewY() )
                        .setDuration(0)
                        .start();
                return true; //Consumed
            case MotionEvent.ACTION_UP:
                float upRawX = motionEvent.getRawX();
                float upRawY = motionEvent.getRawY();

                float upDX = upRawX - binder.getViewModel().getDownRawX();
                float upDY = upRawY - binder.getViewModel().getDownRawY();

                return binder.getViewModel().canClick(binder.floatingActionButtons.constraintLayout, upDX, upDY);
            default:
                return super.onTouchEvent(motionEvent);
        }
    }

    private void showExtendedFAB() {
        binder.floatingActionButtons.extendedFloatingActionButtonOne.show();
        binder.floatingActionButtons.extendedFloatingActionButtonTwo.show();
        binder.floatingActionButtons.extendedFloatingActionButtonThree.show();
        binder.floatingActionButtons.extendedFloatingActionButtonFour.show();
    }

    private void hideExtendedFab() {
        binder.floatingActionButtons.extendedFloatingActionButtonOne.hide();
        binder.floatingActionButtons.extendedFloatingActionButtonTwo.hide();
        binder.floatingActionButtons.extendedFloatingActionButtonThree.hide();
        binder.floatingActionButtons.extendedFloatingActionButtonFour.hide();
    }
}

File: app/src/main/java/com/example/constraintlayoutapp/MainActivity.kt
package com.example.constraintlayoutapp

import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.View
import android.view.View.OnClickListener
import androidx.activity.enableEdgeToEdge
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.vectordrawable.graphics.drawable.Animatable2Compat
import com.example.constraintlayoutapp.databinding.MainBinder
import com.google.android.material.imageview.ShapeableImageView

class MainActivity : AppCompatActivity(), OnClickListener {

    companion object {
        private val TAG : String = MainActivity::class.java.getSimpleName()
    }

    private var binder : MainBinder? = null
    private val viewModel : MainViewModel by viewModels<MainViewModel>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        binder = DataBindingUtil.setContentView(this, R.layout.activity_main);
        binder?.floatingActionButtonWheel?.setOnClickListener(this@MainActivity)
        binder?.floatingActionButtonFlapper?.setOnClickListener(this@MainActivity)
        viewModel.observeFlapper(1, getTheme()).observe(this@MainActivity) { drawable : Drawable? -> binder?.imageViewNorth?.setImageDrawable(drawable) }
        viewModel.observeFlapper(2, getTheme()).observe(this@MainActivity) { drawable : Drawable? -> binder?.imageViewNorthEast?.setImageDrawable(drawable) }
        viewModel.observeFlapper(3, getTheme()).observe(this@MainActivity) { drawable : Drawable? -> binder?.imageViewEast?.setImageDrawable(drawable) }
        viewModel.observeFlapper(4, getTheme()).observe(this@MainActivity) { drawable : Drawable? -> binder?.imageViewSouthEast?.setImageDrawable(drawable) }
        viewModel.observeFlapper(5, getTheme()).observe(this@MainActivity) { drawable : Drawable? -> binder?.imageViewSouth?.setImageDrawable(drawable) }
        viewModel.observeFlapper(6, getTheme()).observe(this@MainActivity) { drawable : Drawable? -> binder?.imageViewSouthWest?.setImageDrawable(drawable) }
        viewModel.observeFlapper(7, getTheme()).observe(this@MainActivity) { drawable : Drawable? -> binder?.imageViewWest?.setImageDrawable(drawable) }
        viewModel.observeFlapper(8, getTheme()).observe(this@MainActivity) { drawable : Drawable? -> binder?.imageViewNorthWest?.setImageDrawable(drawable) }
        viewModel.observeCenter().observe(this@MainActivity) { angle : Float -> binder?.imageViewCenter?.setRotation(angle) }
        viewModel.observeNorthSector().observe(this@MainActivity) { angle : Float -> setImageCircleAngle(binder?.imageViewNorth, angle) }
        viewModel.observeNorthEastSector().observe(this@MainActivity) { angle : Float -> setImageCircleAngle(binder?.imageViewNorthEast, angle) }
        viewModel.observeEastSector().observe(this@MainActivity) { angle : Float -> setImageCircleAngle(binder?.imageViewEast, angle) }
        viewModel.observeSouthEastSector().observe(this@MainActivity) { angle : Float -> setImageCircleAngle(binder?.imageViewSouthEast, angle) }
        viewModel.observeSouthSector().observe(this@MainActivity) { angle : Float -> setImageCircleAngle(binder?.imageViewSouth, angle) }
        viewModel.observeSouthWestSector().observe(this@MainActivity) { angle : Float -> setImageCircleAngle(binder?.imageViewSouthWest, angle) }
        viewModel.observeWestSector().observe(this@MainActivity) { angle : Float -> setImageCircleAngle(binder?.imageViewWest, angle) }
        viewModel.observeNorthWestSector().observe(this@MainActivity) { angle : Float -> setImageCircleAngle(binder?.imageViewNorthWest, angle) }
        viewModel.observeWheel().observe(this@MainActivity, object : Observer<Boolean?> {
            override fun onChanged(isRotate : Boolean?) {
                binder?.floatingActionButtonWheel?.setImageDrawable(viewModel.getAnimatedVectorDrawableCompat(isRotate!!))
                val animatable = binder!!.floatingActionButtonWheel.getDrawable() as Animatable2Compat
                animatable.start()
            }
        })
        viewModel.observeFlapper().observe(this@MainActivity, object : Observer<Boolean?> {
            override fun onChanged(isFlap : Boolean?) {
                binder?.floatingActionButtonFlapper?.setImageDrawable(viewModel.getAnimatedVectorDrawableCompat(isFlap!!))
                val animatable = binder!!.floatingActionButtonFlapper.getDrawable() as Animatable2Compat
                animatable.start()
            }
        })
    }

    private fun setImageCircleAngle(image : ShapeableImageView?, angle : Float) {
        val layoutParams : ConstraintLayout.LayoutParams = image?.getLayoutParams() as ConstraintLayout.LayoutParams
        layoutParams.circleAngle = angle
        image.setLayoutParams(layoutParams)
    }

    override fun onClick(view : View?) {
        if (view?.getId() == binder?.floatingActionButtonWheel?.getId()) {
            viewModel.toggleWheel();
        } else if (view?.getId() == binder?.floatingActionButtonFlapper?.getId()) {
            viewModel.toggleFlapper();
        }
    }
}